src/js/components/dropdown.js

```diff
@@ -0,0 +1,286 @@
+    this.activeDropdown = null;
+
+    // Bind methods
+    this._handleClick = this._handleClick.bind(this);
+    this._handleKeydown = this._handleKeydown.bind(this);
+
+    this.init();
+  }
+
+  /**
+   * Opens a dropdown menu
+   * @param {HTMLButtonElement} toggle - a button element which toggles the dropdown
+   * @param {HTMLDivElement} menu - a div containing the dropdown menu items
+   */
+  openMenu(toggle, menu) {
+    const toggleButton = toggle || this.toggleElement;
+    const menuList = menu || this.menuElement;
+
+    // Return if disabled dropdown is being opened programmatically
+    if (toggleButton.hasAttribute('disabled')) {
+      return;
+    }
+
+    // If the menu was opened by clicking an associated toggle
+    if (toggleButton && toggleButton !== null) {
+      toggleButton.setAttribute('aria-expanded', 'true');
+    }
+
+    // Remove the 'hidden' attribute to show the menu
+    menuList.setAttribute('aria-hidden', 'false');
+
+    this.activeDropdown = this.element;
+  }
+
+  /**
+   * Closes a dropdown menu
+   * @param {HTMLButtonElement} toggle - a button element which toggles the dropdown
+   * @param {HTMLDivElement} menu - a div containing the dropdown menu items
+   */
+  closeMenu(toggle, menu) {
+    const toggleButton = toggle || this.toggleElement;
+    const menuList = menu || this.menuElement;
+
+    // Return if disabled dropdown is being closed programmatically
+    if (toggleButton.hasAttribute('disabled')) {
+      return;
+    }
+
+    if (toggleButton && toggleButton !== undefined) {
+      toggleButton.setAttribute('aria-expanded', 'false');
+    }
+
+    menuList.setAttribute('aria-hidden', 'true');
+  }
+
+  /**
+   * Creates an object of menu items within a dropdown
+   * @param {HTMLDivElement} menu - a div containing the dropdown menu items
+   * @returns {Object} An object of menu item anchor elements
+   */
+  _setUpMenu(menu) {
+    const menuObject = {};
+
+    // Create a real Array of all the focusable elements in the menu
+    const menuFocusables = Array.prototype.slice.call(
+      menu.querySelectorAll(this.focusableElements)
+    );
+
+    // Create a property to hold an array of all focusables
+    menuObject.all = menuFocusables;
+
+    // Create a property with a reference to the first focusable
+    menuObject.first = menuFocusables[0];
+
+    // Create a property with a reference to the last focusable
+    menuObject.last = menuFocusables[menuFocusables.length - 1];
+
+    return menuObject;
+  }
+
+  _handleClick(event) {
+    const dropdown = event.target.closest(this.dropdownAttribute);
+
+    // If click didn't come from within dropdown component, close all open menus
+    if (!dropdown) {
+      this.closeMenu();
+      return;
+    }
+
+    const toggleButton = dropdown.querySelector(this.toggleAttribute);
+    const menuList = dropdown.querySelector(this.menuAttribute);
+
+    // Delegate event to only this instance of the dropdown
+    if (dropdown === this.element) {
```

Comment:
Let's look at cleaning up this logic a bit. There are a lot of nested if statements here which is generally a sign that some of this logic could be extracted/refactored. I would start by thinking about how you can pull out/combine some of these if statements to know for sure that the click didn't come from within the menu and wasn't on the toggle.

For example you could confirm that the click came from within the menu by doing something like this:

```js
if (this.menuElement.contains(event.target)) return;
/**
 *  After this you would know that the click came from within the menu
 * so you can stop here because you don't want to open or close the menu
*/
```

I'm happy to do some pair refactoring on this if you want to schedule some time to get together. Just let me know.